go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -286,44 +316,90 @@ func newTxThrottlerState(topoServer *topo.Server, config *txThrottlerConfig, tar
 
-	result.topologyWatchers = make(
-		[]TopologyWatcherInterface, 0, len(config.healthCheckCells))
-	for _, cell := range config.healthCheckCells {
-		result.topologyWatchers = append(
-			result.topologyWatchers,
+	ctx, cancel := context.WithCancel(context.Background())
+	state.stopHealthCheck = cancel
+	state.initHealthCheckStream(topoServer, target)
+	hcProcessor := state.healthChecksProcessorFactory(topoServer, target, cellsFromTopo)
+	go hcProcessor(ctx)
+
+	return state, nil
+}
+
+func (ts *txThrottlerState) initHealthCheckStream(topoServer *topo.Server, target *querypb.Target) {
+	ts.healthCheck = healthCheckFactory(topoServer, target.Cell, ts.config.healthCheckCells)
+	ts.healthCheckChan = ts.healthCheck.Subscribe()
+	ts.topologyWatchers = make(
+		[]TopologyWatcherInterface, 0, len(ts.config.healthCheckCells))
+	for _, cell := range ts.config.healthCheckCells {
+		ts.topologyWatchers = append(
+			ts.topologyWatchers,
 			topologyWatcherFactory(
 				topoServer,
-				result.healthCheck,
+				ts.healthCheck,
 				cell,
 				target.Keyspace,
 				target.Shard,
 				discovery.DefaultTopologyWatcherRefreshInterval,
-				discovery.DefaultTopoReadConcurrency))
+				discovery.DefaultTopoReadConcurrency),
+		)
 	}
-	return result, nil
 }
 
-func createTxThrottlerHealthCheck(topoServer *topo.Server, config *txThrottlerConfig, result *txThrottlerState, cell string) {
-	ctx, cancel := context.WithCancel(context.Background())
-	result.stopHealthCheck = cancel
-	result.healthCheck = healthCheckFactory(topoServer, cell, config.healthCheckCells)
-	ch := result.healthCheck.Subscribe()
-	go func(ctx context.Context) {
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			case th := <-ch:
-				result.StatsUpdate(th)
+func (ts *txThrottlerState) closeHealthCheckStream() {
+	if ts.healthCheck == nil {
+		return
+	}
+	for _, watcher := range ts.topologyWatchers {
+		watcher.Stop()
+	}
+	ts.topologyWatchers = nil
+	ts.stopHealthCheck()
+	ts.healthCheck.Close()
+}
+
+func (ts *txThrottlerState) healthChecksProcessorFactory(topoServer *topo.Server, target *querypb.Target, cellsFromTopo bool) func(ctx context.Context) {
+	if cellsFromTopo {
+		return func(ctx context.Context) {
+			cellsUpdateTicker := time.NewTicker(topoCellsRefreshInterval)
+			defer cellsUpdateTicker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-cellsUpdateTicker.C:
+					cells, err := fetchKnownCells(topoServer)
+					if err != nil {
+						log.Errorf("txThrottler: failed to fetch cells from topo: %+v", err)
+						continue
+					}
+					if !reflect.DeepEqual(cells, ts.config.healthCheckCells) {
+						log.Info("txThrottler: restarting healthcheck stream due to topology cells update")
+						ts.config.healthCheckCells = cells
+						ts.closeHealthCheckStream()
+						ts.initHealthCheckStream(topoServer, target)
+					}
+				case th := <-ts.healthCheckChan:
+					ts.StatsUpdate(th)
+				}
+			}
+		}
+	} else {
+		return func(ctx context.Context) {
```

Comment:
linting comment: the `if` block issues a `return`. Therefore, remove the `else` clause, and just straight `return`.

Reply:
@shlomi-noach good point, I'll update that 👍 

Any thoughts on [if the ticker loop should be it's own goroutine](https://github.com/vitessio/vitess/pull/12477#discussion_r1241079610) (which would make this "factory" `func` useless)? I'm on the fence 🙇